### ISSUE TITLE
Why is Farbindn learning nothing (returning null)?

### ISSUE THREAD

**Issue by eglassman** [gustavoasoares/refazer#26]
Will try stepping through the call to LearnExtractions.

**Comment by eglassman** [comment]
I think I found it, @gustavoasoares! 

![image](https://user-images.githubusercontent.com/2320194/28141873-5ce75a6c-6713-11e7-8586-8f27683dc674.png)

`new RankingScore(Grammar.Value)` is returning `null`.

**Comment by eglassman** [comment]
I believe the problem is here:

`public RankingScore(Grammar grammar) : base(grammar, "Score", isComplete: true) { }` 

within

```
namespace Tutor
{
    [SuppressMessage("ReSharper", "UnusedMember.Global")]
    public class RankingScore : Feature<double>
```

**Comment by eglassman** [comment]
I think this is the problem, @gustavoasoares:

![image](https://user-images.githubusercontent.com/2320194/28144122-868995a0-671e-11e7-9dab-ee40932f6401.png)

**Comment by eglassman** [comment]
The question is, why is this equality check getting called on two items that aren't even both PythonNodes? Maybe my implementation of `Find` is incorrect!!

![image](https://user-images.githubusercontent.com/2320194/28144244-480b2d10-671f-11e7-831b-97f8dd6c03f8.png)

**Comment by eglassman** [comment]
I feel stuck because I cannot see what is calling Equals on these two objects of different types, one a good looking PythonNode, the other a weird Utils.Bottom symbol object:

![image](https://user-images.githubusercontent.com/2320194/28145546-eff840b4-6727-11e7-8261-c753370ff0f4.png)

**Comment by eglassman** [comment]
@gustavoasoares I am super stuck now. :/

**Comment by eglassman** [comment]
Update: TIL

> The up tack or falsum (⊥, \bot in LaTeX, U+22A5 in Unicode[1]) is a constant symbol used to represent:
> 
> The bottom element in lattice theory
> The bottom type in type theory
> A logical constant denoting "false" in logic

**Comment by gustavoasoares** [comment]
So... It's quite hard to debug this problem because Prose exception is not showing the Stack trace. But after some debugging and tests I was able to fix the learning and just pushed the changes. My guess is: 

The filter was the first production role in the DSL. The witness function of the filter is a built-in WF in PROSE and it accepts a SubSequenceExampleSpec. However, we were passing as input an ExampleSpec. For now, I created an operator Apply, as the first top operator in the DSL, and in its witness function, I convert the ExampleSpec to a SubsequenceExampleSpec, which is passed to the filter. After that, it worked. 

We can also see if by removing this operator and passing a SubseqExampleSpec as input to Farbindn (not sure yet how to do that), it would also work.
